mypy/checker.py

```diff
@@ -216,6 +216,16 @@ def __init__(self, errors: Errors, modules: Dict[str, MypyFile], options: Option
         # for processing module top levels in fine-grained incremental mode.
         self.recurse_into_functions = True
 
+    def reset(self) -> None:
+        """Cleanup stale state that might be left over from a typechecking run.
+
+        This allows us to reuse TypeChecker objects in fine-grained
+        incremental mode.
+        """
+        self.partial_reported.clear()
+        assert self.partial_types == []
+        assert self.deferred_nodes == []
```

Comment:
Maybe also reset `type_map`? We could ensure that the binder and `scope` are clear (or reset them as well). Also, consider clearing `inferred_attribute_types` and `module_refs`. Or maybe it would be easier to create a new `TypeChecker` object (doesn't need to happen in this PR).

Reply:
I got a noticeable (though not particularly big) performance win by reseting the `TypeChecker` instead of creating a new one.